_posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd

```diff
@@ -0,0 +1,283 @@
+
+A solution is to create a wrapper around `oxford_pet_dataset()` that lets us "hook into" the `.getitem()` method, like so:
+
+```{r}
+pet_dataset <- torch::dataset(
+  
+  inherit = oxford_pet_dataset,
+  
+  initialize = function(..., size, normalize = TRUE, augmentation = NULL) {
+    
+    self$augmentation <- augmentation
+    
+    input_transform <- function(x) {
+      x <- x %>%
+        transform_to_tensor() %>%
+        transform_resize(size) 
+      if (isTRUE(normalize)) x <- x %>%
+        transform_normalize(mean = c(0.485, 0.456, 0.406),
+                            std = c(0.229, 0.224, 0.225))
+      x
+    }
+    
+    target_transform <- function(x) {
+      x <- torch_tensor(x, dtype = torch_long())
+      x <- x[newaxis,..]
+      # interpolation = 0 makes sure we still end up with integer classes
+      x <- transform_resize(x, size, interpolation = 0)
+    }
+    
+    self$split <- split
+    
+    super$initialize(
+      ...,
+      transform = input_transform,
+      target_transform = target_transform
+    )
+    
+  },
+  .getitem = function(i) {
+    
+    item <- super$.getitem(i)
+    if (!is.null(self$augmentation)) 
+      self$augmentation(item)
+    else
+      list(x = item$x, y = item$y[1,..])
+  }
+)
+```
+
+All we have to do now is create a custom function that lets us decide on what augmentation to apply to each input-target pair, and then, manually call the respective transformation functions.
+
+Here, we flip, on average, every second image, and if we do, we flip the mask as well. The second transformation -- orchestrating random changes in brightness, saturation, and contrast -- is applied to the input image only.
+
+```{r}
+augmentation <- function(item) {
+  
+  vflip <- runif(1) > 0.5
+  
+  x <- item$x
+  y <- item$y
+  
+  if (isTRUE(vflip)) {
+    x <- transform_vflip(x)
+    y <- transform_vflip(y)
+  }
+  
+  x <- transform_color_jitter(x, brightness = 0.5, saturation = 0.3, contrast = 0.3)
+  
+  list(x = x, y = y[1,..])
+```
+
+We now make use of the wrapper, `pet_dataset()`, to instantiate the training and validation sets, and create the respective data loaders.
+
+```{r}
+train_ds <- pet_dataset(root = dir,
+                        split = "train",
+                        size = c(224, 224),
+                        augmentation = augmentation)
+valid_ds <- pet_dataset(root = dir,
+                        split = "valid",
+                        size = c(224, 224))
+
+train_dl <- dataloader(train_ds, batch_size = 32, shuffle = TRUE)
+valid_dl <- dataloader(valid_ds, batch_size = 32)
+```
+
+### Model training and (visual) evaluation
+
+With `luz`, model training is a matter of two verbs, `setup()` and `fit()`. The learning rate has been determined, for this specific case, using `luz::lr_finder()`; you will likely have to change it when experimenting with different forms of data augmentation (and different data sets).
+
+```{r}
+model <- model %>%
```

Comment:
I think it would be useful to have a reference on to where one can find the model definition.

Reply:
Oops??? I have no idea what happened here ... that model definition part went missing! Normally I keep intermediate backups (not on GH), - will add it back tomorrow ...
What luck you spotted this!!!